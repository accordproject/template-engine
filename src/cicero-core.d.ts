declare module '@accordproject/cicero-core';