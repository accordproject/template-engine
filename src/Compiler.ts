/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import traverse from 'traverse';
import { copySync } from 'fs-extra';

import { TemplateMarkModel, CiceroMarkModel, CommonMarkModel, ConcertoMetaModel } from '@accordproject/markdown-common';

import { ClassDeclaration, Introspector, ModelManager, ModelUtil, Resource } from '@accordproject/concerto-core';
import { CodeGen } from '@accordproject/concerto-tools';
import { FileWriter } from '@accordproject/concerto-util';
import { getCompiler } from './compilers/NodeCompilers';
import { RUNTIME_DIR, writeEpilog, writeImports, writeProlog } from './compilers/Common';

export type ProcessingFunction = (fw:FileWriter, level:number, resource:any) => void;

export type CompilationOptions = {
    skipGenerator?: boolean;
    skipCopyRuntime?: boolean;
}

/**
 * Converts TemplateMark DOM to a Typescript function, that, when run with JSON data
 * that conforms to the template model will produce AgreementMark JSON output.
 */
export class Compiler {
    modelManager: ModelManager;
    templateClass: ClassDeclaration;

    constructor(modelManager: ModelManager) {
        this.modelManager = modelManager;
        const introspector = new Introspector(this.modelManager);
        const templateModels = introspector.getClassDeclarations().filter((item) => {
            return !item.isAbstract() && item.getDecorator('template');
        });
        if (templateModels.length > 1) {
            throw new Error('Found multiple concepts with @template decorator. The model for the template must contain a single concept with the @template decorator.');
        } else if (templateModels.length === 0) {
            throw new Error('Failed to find a concept with the @template decorator. The model for the template must contain a single concept with the @template decoratpr.');
        } else {
            this.templateClass = templateModels[0];
        }
    }

    writeFunction(fw: FileWriter, functionName: string, returnType:string, code: string) {
        fw.writeLine(0, `export function ${functionName}(data:TemplateModel.I${this.templateClass.getName()}, library:any) : ${returnType} {`);
        this.templateClass.getProperties().forEach((p: any) => {
            fw.writeLine(1, `const ${p.getName()} = data.${p.getName()};`);
        });
        fw.writeLine(1, 'const now = dayjs();');
        fw.writeLine(1, code.trim());
        fw.writeLine(0, '}');
        fw.writeLine(0, '');
    }

    /**
     * Get a model manager with templatemark model
     * @param {string} outputDir the output directory
     * @returns {ModelManager} the model manager
     */
    getTemplateMarkModelManager(): ModelManager {
        const modelManager = new ModelManager({ strict: true });
        modelManager.addCTOModel(ConcertoMetaModel.MODEL, 'concertometamodel.cto');
        modelManager.addCTOModel(CommonMarkModel.MODEL, 'commonmark.cto');
        modelManager.addCTOModel(TemplateMarkModel.MODEL, 'templatemark.cto');
        return modelManager;
    }

    /**
     * Get a model manager with templatemark model
     * @param {string} outputDir the output directory
     * @returns {ModelManager} the model manager
     */
    getCiceroMarkModelManager(): ModelManager {
        const modelManager = new ModelManager({ strict: true });
        modelManager.addCTOModel(ConcertoMetaModel.MODEL, 'concertometamodel.cto');
        modelManager.addCTOModel(CommonMarkModel.MODEL, 'commonmark.cto');
        modelManager.addCTOModel(CiceroMarkModel.MODEL, 'ciceromark.cto');
        return modelManager;
    }

    generateTypeScript(modelManager: ModelManager, outputDir: string) {
        // generate the TS code for the template model
        const tsVisitor = new CodeGen.TypescriptVisitor();
        const parameters = {
            fileWriter: new FileWriter(outputDir)
        };
        modelManager.accept(tsVisitor, parameters);
    }

    static doIt(fw:FileWriter, level:number, templateMark: any) {
        fw.writeLine(level, `// start processing ${ModelUtil.getShortName(templateMark.$class)}`);
        traverse(templateMark).forEach(function (context: any) {
            if (typeof context === 'object' && context.$class && typeof context.$class === 'string') {
                const nodeClass = context.$class as string;

                const compiler = getCompiler(nodeClass);
                const sourceLevel = level+1;

                this.before( () => {
                    compiler.enter(fw, sourceLevel, context);
                    compiler.generate(fw, sourceLevel, context, Compiler.doIt);
                });

                this.after( () => {
                    compiler.exit(fw, sourceLevel, context);
                });
            }
        });
        fw.writeLine(level, `// end processing ${ModelUtil.getShortName(templateMark.$class)}`);
    }
    /**
     * Converts a TemplateMark document to typescript code that is parameterised
     * by the 'data' argument.
     *
     * The strategy is to traverse the template mark json, converting each node into
     * typescript statements that add an equivalent ciceromark node. Variables will
     * pull values from 'data'.
     *
     * Conditionals and formulae will call the functions produced by
     * compileUserCode.
     *
     * Note that code generation is itself recursive, for example, when
     * a list definition node is encountered the list must be iterated over
     * based on a variable value coming from data, and code must be
     * generated to produce each list item.
     *
     * '$curNode' is the current ciceromark node that is being populated.
     * Typescript lexical scoping (blocks) is used to introduce local
     * variables.
     *
     * All internal variables are prefixed with '$' to ensure (?)
     * that they cannot collide with end user variable names.
     *
     * @param {IDocument} templateMark the TemplateMark Concerto resource
     * @param {string} outputDir the output directory for code
     */
    compileGenerator(templateMark: Resource, outputDir: string) : void {
        const fw = new FileWriter(outputDir);

        fw.openFile('generator.ts');
        fw.writeLine(0, '// generated code, do not modify');
        fw.writeLine(0, '');

        fw.writeLine(0, '// IMPORTS');
        writeImports(fw, 0, this.templateClass);
        fw.writeLine(0, '');

        fw.writeLine(0, '// GENERATOR');
        writeProlog(fw, 0, this.templateClass);

        Compiler.doIt(fw, 1, templateMark);

        writeEpilog(fw,0);
        fw.closeFile();
    }

    copyRuntime(outputDir: string) {
        copySync('./src/drafting', `${outputDir}/${RUNTIME_DIR}/drafting`);
        copySync('./src/Runtime.ts', `${outputDir}/${RUNTIME_DIR}/Runtime.ts`);
    }

    compile(templateMark: Resource, outputDir: string, options?:CompilationOptions) {
        // generate the model
        this.generateTypeScript(this.modelManager, `${outputDir}`);
        this.generateTypeScript(this.getTemplateMarkModelManager(), `${outputDir}`);
        this.generateTypeScript(this.getCiceroMarkModelManager(), `${outputDir}`);

        // generate names for all the nodes containing user code
        const namedTemplateMark = this.nameUserCode(templateMark);

        // compile the user code
        this.compileUserCode(namedTemplateMark, outputDir);

        // compile the generator
        if(!options?.skipCopyRuntime) {
            this.copyRuntime(outputDir);
        }

        // compile the generator
        if(!options?.skipGenerator) {
            this.compileGenerator(namedTemplateMark, outputDir);
        }
    }

    nameUserCode(templateMark: Resource) {
        const templateMarkDom = templateMark.toJSON();
        return traverse(templateMarkDom).map(function (x) {
            if (x && ((x.$class === `${TemplateMarkModel.NAMESPACE}.ConditionalDefinition` && x.condition) ||
                (`${TemplateMarkModel.NAMESPACE}.ClauseDefinition` && x.condition))) {
                x.functionName = `condition_${this.path.join('_')}`;
            }
            this.update(x);
        });
    }

    compileUserCode(templateMark: Resource, outputDir: string) {
        const fw = new FileWriter(outputDir);

        fw.openFile('usercode.ts');
        fw.writeLine(0, '// generated code, do not modify');
        fw.writeLine(0, '');

        fw.writeLine(0, '// IMPORTS');
        fw.writeLine(0, `import * as TemplateModel from './${this.templateClass.getNamespace()}';`);
        fw.writeLine(0, 'import dayjs from \'dayjs\';');
        fw.writeLine(0, 'import jp from \'jsonpath\';');
        fw.writeLine(0, '');

        const templateMarkDom = templateMark;

        const functionNodes: [] = traverse(templateMarkDom).reduce(function (acc, x) {
            if (x && x.$class === `${TemplateMarkModel.NAMESPACE}.FormulaDefinition`) {
                acc.push({
                    name: x.name,
                    code: x.code,
                });
            }
            return acc;
        }, []);

        const conditionalNodes: [] = traverse(templateMarkDom).reduce(function (acc, x) {
            if (x && x.$class === `${TemplateMarkModel.NAMESPACE}.ConditionalDefinition` && x.condition) {
                acc.push({
                    name: x.functionName,
                    code: x.condition,
                });
            }
            return acc;
        }, []);

        const clauseNodes: [] = traverse(templateMarkDom).reduce(function (acc, x) {
            if (x && x.$class === `${TemplateMarkModel.NAMESPACE}.ClauseDefinition` && x.condition) {
                acc.push({
                    name: x.functionName,
                    code: x.condition,
                });
            }
            return acc;
        }, []);

        fw.writeLine(0, '// FORMULAE');
        functionNodes.forEach((fun: any) => {
            this.writeFunction(fw, `${fun.name}`, 'any', fun.code.contents);
        });
        fw.writeLine(0, '');

        fw.writeLine(0, '// CONDITIONALS');
        conditionalNodes.forEach((fun: any) => {
            this.writeFunction(fw, `${fun.name}`, 'boolean', `return ${fun.code.contents}`);
        });
        fw.writeLine(0, '');

        fw.writeLine(0, '// CLAUSES');
        clauseNodes.forEach((fun: any) => {
            this.writeFunction(fw, `${fun.name}`, 'boolean', `return ${fun.code.contents}`);
        });
        fw.writeLine(0, '');

        fw.closeFile();
    }
}