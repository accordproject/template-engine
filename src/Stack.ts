/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
interface IStack<T> {
    push(item: T): void;
    pop(): T | undefined;
    peek(): T | undefined;
    size(): number;
  }

class Stack<T> implements IStack<T> {
    private storage: T[] = [];

    constructor(private capacity: number = Infinity) {}

    push(item: T): void {
        if (this.size() === this.capacity) {
            throw Error('Stack has reached max capacity, you cannot add more items');
        }
        this.storage.push(item);
    }

    pop(): T | undefined {
        return this.storage.pop();
    }

    peek(): T | undefined {
        return this.storage[this.size() - 1];
    }

    size(): number {
        return this.storage.length;
    }
}