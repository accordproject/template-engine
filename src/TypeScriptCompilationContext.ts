/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { ClassDeclaration, ModelManager } from '@accordproject/concerto-core';
import { CodeGen } from '@accordproject/concerto-codegen';
import { InMemoryWriter } from '@accordproject/concerto-util';
import { getTemplateClassDeclaration } from './utils';

/**
 * This class creates the typescript types
 * required to compile Typescript expressions (used in
 * formulae, conditions and clauses) to JavaScript. It uses
 * these to create a compilation context for '@typescript/twoslash'
 * which is used to compile the typescript code.
 */
export class TypeScriptCompilationContext {

    modelManager:ModelManager;
    templateClass:ClassDeclaration;

    constructor(modelManager:ModelManager,templateConceptFqn?: string) {
        this.modelManager = modelManager;
        this.templateClass = getTemplateClassDeclaration(this.modelManager, templateConceptFqn);
    }

    getTypeScriptFiles() : Record<string,string> {
        const result:Record<string,string> = {};

        const visitor = new CodeGen.TypescriptVisitor();
        const writer = new InMemoryWriter();

        const params = {
            fileWriter: writer
        };
        this.modelManager.accept(visitor, params);
        writer.getFilesInMemory().forEach( (value, key) => {
            result[key] = value;
        });
        return result;
    }

    getCompilationContext() : string {
        const files = this.getTypeScriptFiles();

        let result = '';

        Object.keys(files).forEach( key => {
            const content = files[key];
            result += `
// @filename: ${key}
${content}
`;
        });

        result += `
// @filename: code.ts
import * as TemplateModel from './${this.templateClass.getNamespace()}';
import dayjs from 'dayjs';
import jp from 'jsonpath';

type GenerationOptions = {
    now?:string,
    locale?:string
}
`;
        return result;
    }
}