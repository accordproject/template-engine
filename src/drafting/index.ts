/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import booleanDrafter from './Boolean';
import dateTimeDrafter from './DateTime';
import doubleDrafter from './Double';
import integerDrafter from './Integer';
import longDrafter from './Long';
import monetaryAmountDrafter from './MonetaryAmount';
import stringDrafter from './String';

export function getDrafter(typeName: string) : ((value:any, format:string) => string)|null  {
    switch(typeName) {
    case 'Boolean': return booleanDrafter;
    case 'DateTime': return dateTimeDrafter;
    case 'Double': return doubleDrafter;
    case 'Integer': return integerDrafter;
    case 'Long': return longDrafter;
    case 'org.accordproject.money@0.3.0.MonetaryAmount': return monetaryAmountDrafter;
    case 'String': return stringDrafter;
    default: return null;
    }
}