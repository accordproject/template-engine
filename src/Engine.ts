/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import dayjs from 'dayjs';
import jp from 'jsonpath';

import traverse from 'traverse';

import { ClassDeclaration, Factory, Introspector, ModelManager, Serializer } from '@accordproject/concerto-core';
import { draftingMap } from './drafting';
import { TemplateMarkModel } from './externalModels/TemplateMarkModel';
import { CommonMarkModel } from './externalModels/CommonMarkModel';
import { AgreementMarkModel } from './externalModels/AgreementMarkModel';
import { ConcertoMetaModel } from './externalModels/ConcertoMetaModel';

// used to migrate old template mark json to latest namespaces
const TEMPLATEMARK_OLD_RE = /^(org\.accordproject\.templatemark)\.(\w+)$/;
const COMMONMARK_OLD_RE = /^(org\.accordproject\.commonmark)\.(\w+)$/;
const CONCERTOMETAMODEL_OLD_RE = /^(concerto\.metamodel)\.(\w+)$/;

// use to create agreementmark from templatemark
const TEMPLATEMARK_RE = /^(org\.accordproject\.templatemark)@(.+)\.(\w+)Definition$/;
const FORMULA_DEFINITION_RE = /^(org\.accordproject\.templatemark)@(.+)\.FormulaDefinition$/;
const VARIABLE_DEFINITION_RE = /^(org\.accordproject\.templatemark)@(.+)\.VariableDefinition$/;
const CONDITIONAL_DEFINITION_RE = /^(org\.accordproject\.templatemark)@(.+)\.ConditionalDefinition$/;
const ENUM_VARIABLE_DEFINITION_RE = /^(org\.accordproject\.templatemark)@(.+)\.EnumVariableDefinition$/;
const FORMATTED_VARIABLE_DEFINITION_RE = /^(org\.accordproject\.templatemark)@(.+)\.FormattedVariableDefinition$/;
const WITH_DEFINITION_RE = /^(org\.accordproject\.templatemark)@(.+)\.WithDefinition$/;
const LISTBLOCK_DEFINITION_RE = /^(org\.accordproject\.templatemark)@(.+)\.ListBlockDefinition$/;
const JOIN_DEFINITION_RE = /^(org\.accordproject\.templatemark)@(.+)\.JoinDefinition$/;
const OPTIONAL_DEFINITION_RE = /^(org\.accordproject\.templatemark)@(.+)\.OptionalDefinition$/;

// TODO??
// ContractDefinition
// ClauseDefinition
// OptionalDefinition
// Document

type TemplateData = Record<string, unknown>;

/**
 * Evaluates a JS expression
 * @param {*} data the contract data
 * @param {string} expression the JS expression
 * @param {Date} now the current value for now
 * @returns {object} the result of evaluating the expression against the data
 */
function evaluateJavaScript(data: TemplateData, expression: string, now?: dayjs.Dayjs): object {
    if (!data || !expression) {
        throw new Error(`Cannot evaluate JS ${expression} against ${data}`);
    }
    data.now = now ? now : dayjs();
    const args = Object.keys(data);
    const values = Object.values(data);
    const types = values.map(v => typeof v);
    const DEBUG = false;
    if (DEBUG) {
        console.debug('**** ' + JSON.stringify(data, null, 2));
        console.debug('**** ' + expression);
        console.debug('**** ' + args);
        console.debug('**** ' + values);
        console.debug('**** ' + types);
    }
    const fun = new Function(...args, expression); // SECURITY!
    const result = fun(...values);
    if (DEBUG) {
        console.debug('**** ' + result);
    }
    return result;
}

/**
 * Calculates a JSON path to use to retrieve data, based on a TemplatemMark tree.
 * For example, if we hit a VariableDefinition {{city}} that is nested inside a
 * WithDefinition {{#with address}} then the JSON path returned should be '$.address.city'.
 * Similarly ListBlockDefinition and JoinDefinition also include property names that must
 * apply to their child nodes.
 * @param {*} rootData the root of the JSON document, typically this is a TemplateMark JSON
 * @param {*} currentNode the current TemplateMark node we are processing
 * @param {string[]} paths the traverse path to the current node
 * @returns {string} the JSON path to use to retrieve data
 */
function getJsonPath(rootData:any, currentNode:any, paths:string[]) : string {
    if(!currentNode) {
        throw new Error('Data must be supplied');
    }
    if(!currentNode.name) {
        throw new Error(`Data must have a name: ${JSON.stringify(currentNode)}`);
    }
    if(currentNode.name === 'this') { // TODO: is this what we want to do?!
        return '$.';
    }
    if(currentNode.name.indexOf('.') >=0) {
        // prevent JSON path injection
        throw new Error(`Invalid name property ${currentNode.name}`);
    }
    if(!paths || !paths.length || paths.length <1 ) {
        throw new Error('Paths must be supplied');
    }
    const withPath = [];
    for (let n = 1; n < paths.length; n++) {
        const sub = paths.slice(0, n);
        const obj = traverse.get(rootData, sub);
        if(!obj) {
            throw new Error(`Failed to find data with path ${sub} and data ${JSON.stringify(rootData)}`);
        }
        if (obj.$class) {
            if(obj.$class === 'org.accordproject.templatemark@1.0.0.ListBlockDefinition' ||
            obj.$class === 'org.accordproject.templatemark@1.0.0.WithDefinition' ||
            obj.$class === 'org.accordproject.templatemark@1.0.0.JoinDefinition' ||
            obj.$class === 'org.accordproject.templatemark@1.0.0.OptionalDefinition') {
                withPath.push(obj.name);
            }
        }
    }
    const path = withPath.length > 0 ? `${withPath.join('.')}.${currentNode.name}` : currentNode.name;
    return `$.${path}`;
}

/**
 * Generates an AgreementMark JSON document from a template plus data.
 * @param {*} templateMark - the TemplateMark JSON document
 * @param {*} data - the template data JSON
 * @returns {*} the AgreementMark JSON
 */
function generateAgreement(templateMark: object, data: TemplateData): any {
    return traverse(templateMark).map(function (context: any) {
        let stopHere = false;
        if (typeof context === 'object' && context.$class && typeof context.$class === 'string') {
            const nodeClass = context.$class as string;

            // rewrite node types, mapping from TemplateMark to AgreementMark
            const match = nodeClass.match(TEMPLATEMARK_RE);
            if (match && match.length > 1) {
                context.$class = `org.accordproject.agreementmark@${match[2]}.${match[3]}`;
            }

            // convert a WithDefinition to a Paragraph in the output
            // not 100% sure we want to do that ... we may need to process
            // the child variable nodes and reparent them?
            if (WITH_DEFINITION_RE.test(nodeClass)) {
                context.$class = 'org.accordproject.commonmark@1.0.0.Paragraph';
                delete context.name;
                delete context.elementType;
            }

            // add a 'value' property to FormulaDefinition
            // with the result of evaluating the JS code
            else if (FORMULA_DEFINITION_RE.test(nodeClass)) {
                if (context.code) {
                    context.value = JSON.stringify(evaluateJavaScript(data, context.code));
                }
                else {
                    throw new Error('Formula node is missing code.');
                }
            }

            else if (LISTBLOCK_DEFINITION_RE.test(nodeClass)) {
                const path = getJsonPath(templateMark, context, this.path);
                const variableValues = jp.query(data, path, 1);

                if (variableValues.length === 0) {
                    throw new Error(`No values found for path '${path}' in data ${data}.`);
                }
                else {
                    const arrayData = variableValues[0];
                    if(!Array.isArray(arrayData)) {
                        throw new Error(`Values found for path '${path}' in data ${data} is not an array: ${arrayData}.`);
                    }
                    else {
                        context.$class = 'org.accordproject.commonmark@1.0.0.List';
                        delete context.elementType;
                        delete context.name;
                        context.nodes = arrayData.map( arrayItem => {
                            // arrayItem is now the data for the nested traversal
                            const subResult = generateAgreement(context.nodes[0].nodes[0], arrayItem);
                            return {
                                $class: 'org.accordproject.commonmark@1.0.0.Item',
                                nodes: subResult.nodes ? subResult.nodes : []
                            };
                        });
                        stopHere = true; // do not process child nodes, we've already done it above...
                    }
                }
            }

            else if (JOIN_DEFINITION_RE.test(nodeClass)) {
                const path = getJsonPath(templateMark, context, this.path);
                const variableValues = jp.query(data, path, 1);

                if (variableValues.length === 0) {
                    throw new Error(`No values found for path '${path}' in data ${data}.`);
                }
                else {
                    const arrayData = variableValues[0];
                    if(!Array.isArray(arrayData)) {
                        throw new Error(`Values found for path '${path}' in data ${data} is not an array: ${arrayData}.`);
                    }
                    else {
                        context.$class = 'org.accordproject.commonmark@1.0.0.Text';
                        const drafter = draftingMap.get(context.elementType);
                        context.text = arrayData.map( arrayItem => {
                            return drafter ? drafter(arrayItem, context.format) : arrayItem as string;
                        }).join(context.separator);
                        delete context.elementType;
                        delete context.name;
                        delete context.separator;
                        delete context.nodes;
                        stopHere = true; // do not process child nodes, we've already done it above...
                    }
                }
            }

            // add a 'value' property to VariableDefinition
            // with the value of the variable from 'data'
            else if (VARIABLE_DEFINITION_RE.test(nodeClass) ||
                ENUM_VARIABLE_DEFINITION_RE.test(nodeClass) ||
                FORMATTED_VARIABLE_DEFINITION_RE.test(nodeClass)) {
                const path = getJsonPath(templateMark, context, this.path);
                const variableValues = jp.query(data, path, 1);

                if (variableValues.length === 0) {
                    throw new Error(`No values found for path '${path}' in data ${data}.`);
                }
                else {
                    // convert the value to a string, optionally using the formatter
                    const variableValue = variableValues[0];
                    const drafter = draftingMap.get(context.elementType);
                    context.value = drafter ? drafter(variableValue, context.format) : variableValue as string;
                }
            }

            // add a 'isTrue' property to ConditionDefinition
            // with the result of evaluating the JS code
            else if (CONDITIONAL_DEFINITION_RE.test(nodeClass)) {
                if (context.condition) {
                    context.isTrue = evaluateJavaScript(data, `return !!${context.condition}`) as unknown as boolean;
                    if(context.isTrue) {
                        delete context.whenFalse;
                    }
                    else {
                        delete context.whenTrue;
                    }
                }
                else {
                    throw new Error('Condition node is missing condition.');
                }
            }

            // add a 'hasSome' property to OptionalDefinition
            else if (OPTIONAL_DEFINITION_RE.test(nodeClass)) {
                const path = getJsonPath(templateMark, context, this.path);
                const variableValues = jp.query(data, path, 1);
                if (variableValues && variableValues.length) {
                    if(variableValues.length === 1) {
                        context.hasSome = !!variableValues[0];
                        delete context.whenNone;
                    }
                    else {
                        throw new Error(`Multiple values found for path '${path}' in data ${data}.`);
                    }
                }
                else {
                    context.hasSome = false;
                    delete context.whenSome;
                }
            }
        }
        this.update(context, stopHere);
    });
}

/**
 * Migrates a TemplateMark JSON document without namespace versions
 * to a document with namespace versions.
 * @param {*} templateMark - the TemplateMark JSON document
 * @returns {*} the TemplateMark JSON with namespace versions
 */
function migrateTemplateMark(templateMark: object): object {
    return traverse(templateMark).map(function (x: any) {
        if (typeof x === 'object' && x.$class && typeof x.$class === 'string') {
            const nodeClass = x.$class as string;
            {
                const match = nodeClass.match(TEMPLATEMARK_OLD_RE);
                if (match && match.length > 1) {
                    x.$class = `org.accordproject.templatemark@1.0.0.${match[2]}`;
                }
            }
            {
                const match = nodeClass.match(COMMONMARK_OLD_RE);
                if (match && match.length > 1) {
                    x.$class = `org.accordproject.commonmark@1.0.0.${match[2]}`;
                }
            }
            {
                const match = nodeClass.match(CONCERTOMETAMODEL_OLD_RE);
                if (match && match.length > 1) {
                    x.$class = `concerto.metamodel@1.0.0.${match[2]}`;
                }
            }
        }
        this.update(x);
    });
}

/**
 * A template engine: merges the markup and logic of a template with
 * JSON data to produce JSON data.
 */
export class Engine {
    modelManager: ModelManager;
    templateClass: ClassDeclaration;

    constructor(modelManager: ModelManager) {
        this.modelManager = modelManager;
        const introspector = new Introspector(this.modelManager);
        const templateModels = introspector.getClassDeclarations().filter((item) => {
            return !item.isAbstract() && item.getDecorator('template');
        });
        if (templateModels.length > 1) {
            throw new Error('Found multiple concepts with @template decorator. The model for the template must contain a single concept with the @template decorator.');
        } else if (templateModels.length === 0) {
            throw new Error('Failed to find a concept with the @template decorator. The model for the template must contain a single concept with the @template decoratpr.');
        } else {
            this.templateClass = templateModels[0];
        }
    }

    /**
     * Migrates a template mark JSON document
     * @param {*} templateMark the TemplateMark JSON object
     * @returns {*} TemplateMark JSON migrated to latest namespace version
     */
    migrate(templateMark: object): object {
        return migrateTemplateMark(templateMark);
    }

    /**
     * Checks that a TemplateMark JSON document is valid with respect to the
     * TemplateMark model, as well as the template model.
     *
     * Checks:
     * 1. Variable names are valid properties in the template model
     * 2. Optional properties have guards
     * @param {*} templateMark the TemplateMark JSON object
     * @param {*} templateData the agreement data
     * @returns {*} TemplateMark JSON that has been typed checked and has type metadata added
     * @throws {Error} if the templateMark document is invalid
     */
    checkTypes(templateMark: object, templateData: object): object {
        const modelManager = new ModelManager({ strict: true });
        modelManager.addCTOModel(ConcertoMetaModel, 'concertometamodel.cto');
        modelManager.addCTOModel(CommonMarkModel, 'commonmark.cto');
        modelManager.addCTOModel(TemplateMarkModel, 'templatemark.cto');
        const factory = new Factory(modelManager);
        const serializer = new Serializer(factory, modelManager);
        serializer.fromJSON(templateMark);
        return templateMark;
    }

    validateCiceroMark(ciceroMark: object) {
        const modelManager = new ModelManager({ strict: true });
        modelManager.addCTOModel(ConcertoMetaModel, 'concertometamodel.cto');
        modelManager.addCTOModel(CommonMarkModel, 'commonmark.cto');
        modelManager.addCTOModel(AgreementMarkModel, 'agreementmark.cto');
        const factory = new Factory(modelManager);
        const serializer = new Serializer(factory, modelManager);
        serializer.fromJSON(ciceroMark);
        return true;
    }

    generate(templateMark: object, data: TemplateData): any {
        const factory = new Factory(this.modelManager);
        const serializer = new Serializer(factory, this.modelManager);
        const templateData = serializer.fromJSON(data);
        if (templateData.getFullyQualifiedType() !== this.templateClass.getFullyQualifiedName()) {
            throw new Error(`Template data must be of type '${this.templateClass.getFullyQualifiedName()}'.`);
        }
        const typedTemplateMark = this.checkTypes(templateMark, data);
        const ciceroMark = generateAgreement(typedTemplateMark, data);
        this.validateCiceroMark(ciceroMark);
        return ciceroMark;
    }
}