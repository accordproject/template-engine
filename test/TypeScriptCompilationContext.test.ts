import { ModelManager } from '@accordproject/concerto-core';
import { readFileSync } from 'fs-extra';
import { TypeScriptCompilationContext } from '../src/TypeScriptCompilationContext';

describe('typescript compilation context', () => {
    test('should create a context from a model manager', async () => {
        const model = readFileSync('./test/templates/good/full/model.cto', 'utf-8');
        const modelManager = new ModelManager();
        modelManager.addCTOModel(model, 'model.cto');
        const context = new TypeScriptCompilationContext(modelManager);
        const result = context.getCompilationContext();
        expect(result).toMatchSnapshot();
    });
});