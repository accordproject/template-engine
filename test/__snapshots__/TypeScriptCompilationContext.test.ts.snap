// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typescript compilation context should create a context from a model manager 1`] = `
"
// @filename: concerto@1.0.0.ts
/* eslint-disable @typescript-eslint/no-empty-interface */
// Generated code for namespace: concerto@1.0.0

// imports

// Warning: Beware of circular dependencies when modifying these imports
import type {
	IAddress,
	IOrder,
	ILoyaltyStatus,
	IPreferences,
	ITemplateData
} from './test@1.0.0';

// interfaces
export interface IConcept {
   $class: string;
}

export type ConceptUnion = IAddress | 
IOrder | 
ILoyaltyStatus | 
IPreferences | 
ITemplateData;

export interface IAsset extends IConcept {
   $identifier: string;
}

export interface IParticipant extends IConcept {
   $identifier: string;
}

export interface ITransaction extends IConcept {
   $timestamp: Date;
}

export interface IEvent extends IConcept {
   $timestamp: Date;
}



// @filename: concerto.ts
/* eslint-disable @typescript-eslint/no-empty-interface */
// Generated code for namespace: concerto

// imports

// interfaces
export interface IConcept {
   $class: string;
}

export interface IAsset extends IConcept {
   $identifier: string;
}

export interface IParticipant extends IConcept {
   $identifier: string;
}

export interface ITransaction extends IConcept {
}

export interface IEvent extends IConcept {
}



// @filename: test@1.0.0.ts
/* eslint-disable @typescript-eslint/no-empty-interface */
// Generated code for namespace: test@1.0.0

// imports
import {IConcept} from './concerto@1.0.0';

// interfaces
export interface IAddress extends IConcept {
   street: string;
   city: string;
   zip: string;
}

export interface IOrder extends IConcept {
   sku: string;
   amount: number;
}

export interface ILoyaltyStatus extends IConcept {
   level: string;
}

export interface IPreferences extends IConcept {
   favoriteColors?: string[];
}

export interface ITemplateData extends IConcept {
   firstName: string;
   lastName: string;
   middleNames: string[];
   active: boolean;
   lastVisit: Date;
   address: IAddress;
   orders: IOrder[];
   loyaltyStatus?: ILoyaltyStatus;
   preferences: IPreferences;
}



// @filename: code.ts
import * as TemplateModel from './test@1.0.0';
import dayjs from 'dayjs';
import jp from 'jsonpath';
"
`;
